csi/node_server.go

```diff
@@ -479,15 +489,25 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 		return nil, status.Error(codes.InvalidArgument, "volume id missing in request")
 	}
 
+	mounter := mount.New("")
+
 	// CO owns the staging_path so we only unmount but not remove the path
-	if err := unmount(targetPath, mount.New("")); err != nil {
+	if err := unmount(targetPath, mounter); err != nil {
 		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to unmount volume %s mount point %v error %v", volumeID, targetPath, err))
 	}
 
+	// for block mode we use the staging path as parent, so we have to do additional cleanup of the subfolder/files
+	// we should transition the regular fs mounts to also use the same sub folder, this allows us to store additional
+	// metadata as well as do more forcefull removals since we no longer share the control of the staging_path with kubernetes
+	deviceFilePath := filepath.Join(targetPath, volumeID)
+	if err := cleanupMountPoint(deviceFilePath, mounter); err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to clean up volume %s device mount point %v error %v", volumeID, deviceFilePath, err))
+	}
+
```

Comment:
Suggest doing this only for block volume

Should we do this unmount operation for the staged target mounted path before the above unmounting of the published target mounted path? For the nested mount paths, it should be better to unmount the inner first, then outer. WDYT?

Reply:
The same concern:
Shouldn't we clean up the subfolder/files for the staging directory/path first (before unmounting the staging path)?

Reply:
1. I agree I would have preferred to do this only for block mode unfortunately that information is not available as part of the `NodeUnstageVolume` call. Similiar issue as with the NodeExpandVolume for the encrypted device, this is the reason why I recommed we also transition to utilizing a subfolder of the staging path for the regular fs mode, since we can store additional metadata like access type / mode etc :)

2. this is on purpose, since the unmount of the parent is a no op for block mode, this is also important for backwards compatibility of the existing block devices. 

The cleanup of the child file/folder is a no op for fs mode, we can adjust this once we have switched to using a subfolder/file instead of the staging path directly.
```
	// CO owns the staging_path so we only unmount but not remove the path
	err := unmount(targetPath, mounter);

        // subfolder/file that we have full control over
	deviceFilePath := filepath.Join(targetPath, volumeID)
	cleanupMountPoint(deviceFilePath, mounter)
```

Reply:
Got it. Can you add this to the comment?
> the unmount of the parent is a no op for block mode, this is also important for backwards compatibility of the existing block devices.